File: src/walkway_navigation/src/stuck_monitor.cpp
#include "rclcpp/rclcpp.hpp"
#include "action_msgs/msg/goal_status_array.hpp"
#include "nav2_msgs/srv/clear_entire_costmap.hpp"

using namespace std::chrono_literals;
using std::placeholders::_1;

class StuckMonitor : public rclcpp::Node {
    public:
    StuckMonitor() : Node("stuck_monitor"), is_clearing_active_(false) {
        status_sub_ = this->create_subscription<action_msgs::msg::GoalStatusArray>(
            "/navigate_to_pose/_action/status", 
            10,
            std::bind(&StuckMonitor::status_callback, this, _1)
        );

        clear_client_ = this->create_client<nav2_msgs::srv::ClearEntireCostmap>(
            "/local_costmap/clear_entirely_local_costmap"
        );
    }

    private:
    void status_callback(const action_msgs::msg::GoalStatusArray::SharedPtr msg) {
        if (msg->status_list.empty()) return;

        if (msg->status_list.back().status == action_msgs::msg::GoalStatus::STATUS_ABORTED) {
            if (!is_clearing_active_) {
                RCLCPP_WARN(this->get_logger(), "Robot stuck! Triggering clear...");
                trigger_clear();
            }
        }
    }

    void trigger_clear() {
        if (clear_client_->service_is_ready()) {
            is_clearing_active_ = true;

            auto request = std::make_shared<nav2_msgs::srv::ClearEntireCostmap::Request>();

            clear_client_->async_send_request(
                request,
                std::bind(&StuckMonitor::clear_finished_callback, this, std::placeholders::_1)
            );
        }
    }

    void clear_finished_callback(rclcpp::Client<nav2_msgs::srv::ClearEntireCostmap>::SharedFuture future) {
        RCLCPP_INFO(this->get_logger(), "Costmap successfully cleared.");
        is_clearing_active_ = false;
    }

    rclcpp::Subscription<action_msgs::msg::GoalStatusArray>::SharedPtr status_sub_;
    rclcpp::Client<nav2_msgs::srv::ClearEntireCostmap>::SharedPtr clear_client_;
    bool is_clearing_active_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<StuckMonitor>());
    rclcpp::shutdown();
    return 0;
}

File: src/walkway_navigation/scripts/stuck_monitor.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from action_msgs.msg import GoalStatusArray, GoalStatus
from nav2_msgs.srv import ClearEntireCostmap
import time

class StuckMonitor(Node):
    def __init__(self):
        super().__init__('stuck_monitor')

        # 1. Subscribe to the Action Status Topic
        self.status_sub = self.create_subscription(
            GoalStatusArray,
            '/navigate_to_pose/_action/status',
            self.status_callback,
            10
        )

        # 2. Setup the Service Client
        self.local_client = self.create_client(
            ClearEntireCostmap,
            '/local_costmap/clear_entirely_local_costmap'
        )

        self.global_client = self.create_client(
            ClearEntireCostmap,
            '/global_costmap/clear_entirely_global_costmap'
        )

        self.last_clear_time = 0.0
        self.cooldown_duration = 5.0

        #self.is_clearing_active = False

    def status_callback(self, msg):
        if not msg.status_list:
            return
        
        # Get the status of the most recent goal
        if msg.status_list[-1].status == GoalStatus.STATUS_ABORTED:
            current_time = time.time()

            # Check if the cooldown has passed
            if (current_time - self.last_clear_time) > self.cooldown_duration:
                self.get_logger().warn('Robot stuck! Triggering dual costmap clear...')
                self.trigger_clear_costmaps()
                self.last_clear_time = current_time
            else:
                self.get_logger().debug('Aborted status received, but clear is on cooldown.')

    def trigger_clear_costmaps(self):
        req = ClearEntireCostmap.Request()
        if self.local_client.service_is_ready():
            local_future = self.local_client.call_async(req)
        if self.global_client.service_is_ready():
            global_future = self.global_client.call_async(req)

        local_future.add_done_callback(self.clear_finished_callback)

    def clear_finished_callback(self, local_future):
        self.get_logger().info('Costmaps successfully cleared.')

def main(args=None):
    rclpy.init(args=args)
    node = StuckMonitor()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()